modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,451 @@
+			return;
+		}
+
+		$body = json_decode( $response['body'] );
+		$connect_URL = $body->services->instagram->connect_URL;
+		$query_params = $this->get_query_params();
+		$url = add_query_arg(
+			$query_params,
+			$connect_URL
+		);
+		return $url;
+	}
+
+	private function is_legacy_token( $token_id ) {
+		$site = Jetpack_Options::get_option( 'id' );
+		$path = sprintf( '/sites/%s/instagram/%s/check-legacy', $site, $token_id );
+		$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			do_action( 'wpcomsh_log', 'Instagram widget: failed to verify if token is for legacy API: API returned code ' . $response_code );
+			return 'ERROR';
+		}
+		$body = json_decode( $result['body'] );
+		return $body->legacy;
+	}
+
+	/**
+	 * Outputs the widget configuration form for the widget administration page.
+	 * Allows the user to add new Instagram Keyring tokens and more.
+	 *
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function form( $instance ) {
+		$instance = wp_parse_args( $instance, $this->defaults );
+
+		// If coming back to the widgets page from an action, expand this widget
+		if ( isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number ) {
+			echo '<script type="text/javascript">jQuery(document).ready(function($){ $(\'.widget[id$="wpcom_instagram_widget-' . esc_js( $this->number ) . '"] .widget-inside\').slideDown(\'fast\'); });</script>';
+		}
+
+		// If removing the widget's stored token ID
+		elseif ( $instance['token_id'] && isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number && ! empty( $_GET['instagram_widget'] ) && 'remove_token' == $_GET['instagram_widget'] ) {
+			if ( empty( $_GET['nonce'] ) || ! wp_verify_nonce( $_GET['nonce'], 'instagram-widget-remove-token-' . $this->number . '-' . $instance['token_id'] ) ) {
+				wp_die( __( 'Missing or invalid security nonce.', 'wpcom-instagram-widget' ) );
+			}
+			$site = Jetpack_Options::get_option( 'id' );
+			$path = sprintf( '/sites/%s/instagram/%s/remove', $site, $instance['token_id'] );
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+			$response_code = wp_remote_retrieve_response_code( $result );
+
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to remove keyring token: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$instance['token_id'] = $this->defaults['token_id'];
+
+			$this->update_widget_token_id( $instance['token_id'] );
+			$this->update_widget_token_legacy_status( false );
```

Comment:
Once the user removes the token at least once we can assume from that point that it is not a legacy token